samples/nrf_desktop/src/modules/config.c

```diff
@@ -0,0 +1,231 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <stdlib.h>
+
+#include <zephyr.h>
+#include <settings/settings.h>
+
+#include "event_manager.h"
+#include "config_event.h"
+
+#define MODULE config
+#include "module_state_event.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_CONFIG_LOG_LEVEL);
+
+#define KEY_LEN 9
+
+/* Array of pointers to data loaded from settings */
+static struct config_event *loaded_data[CONFIG_EVENT_ID_MAX];
+
+static bool config_id_is_supported(u8_t id)
+{
+	if (IS_ENABLED(CONFIG_DESKTOP_MOTION_OPTICAL_ENABLE)) {
+		if ((id >= CONFIG_EVENT_ID_MOTION_CPI) &&
+		    (id <= CONFIG_EVENT_ID_MOTION_DOWNSHIFT_REST2)) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+static int settings_set(int argc, char **argv, void *val_ctx)
+{
+	if (argc != 1) {
+		return -ENOENT;
+	}
+
+	char *end;
+	u8_t id = strtol(argv[0], &end, 10);
+	if (*end != '\0') {
+		LOG_WRN("Loaded config ID is not a valid number");
+		return 0;
+	}
+
+	if (!config_id_is_supported(id)) {
+		LOG_WRN("ID %u not supported", id);
+		return 0;
+	}
+
+	if (!loaded_data[id]) {
+		loaded_data[id] = new_config_event();
+		loaded_data[id]->id = id;
+		loaded_data[id]->store_needed = false;
+	}
+
+	int len = settings_val_read_cb(val_ctx, loaded_data[id]->data,
+				       sizeof(loaded_data[id]->data));
+	if (len < 0) {
+		LOG_ERR("Can't read config (id:%u err:%d)", id, len);
+		return len;
+	}
+
+	return 0;
+}
+
+static int commit(void)
+{
+	/* All settings loaded, including FCB duplicates.
+	 * We can apply values to application.
+	 */
+	for (u8_t id = 0; id < CONFIG_EVENT_ID_MAX; id++) {
+		if (!loaded_data[id]) {
+			continue;
+		}
+
+		EVENT_SUBMIT(loaded_data[id]);
```

Comment:
add `loaded_data[id] = NULL;`

Reply:
if this is a one time only operation maybe we can allocate `loaded_data` on heap to avoid RAM wasting?

Reply:
Fixed